src/contracts/services/batchQuery.ts

```diff
@@ -42,59 +47,125 @@ function parseBatchQuery(response: BatchQueryResponse): BatchQueryParsedResponse
   });
 }
 
+// Function to divide an array of queries into batches of arrays
+function divideSingleBatchIntoArrayOfMultipleBatches(array: BatchQueryParams[], batchSize: number) {
+  const batches = [];
+  for (let i = 0; i < array.length; i += batchSize) {
+    batches.push(array.slice(i, i + batchSize));
+  }
+  return batches;
+}
+
 /**
  * batch query of multiple contracts/message at a time
  */
-const batchQuery$ = ({
+const batchQuerySingleBatch$ = ({
   contractAddress,
   codeHash,
-  lcdEndpoint,
-  chainId,
   queries,
+  client,
 }:{
   contractAddress: string,
   codeHash?: string,
   lcdEndpoint?: string,
   chainId?: string,
-  queries: BatchQueryParams[]
-}) => getActiveQueryClient$(lcdEndpoint, chainId).pipe(
-  switchMap(({ client }) => sendSecretClientContractQuery$({
-    queryMsg: msgBatchQuery(queries),
-    client,
-    contractAddress,
-    codeHash,
-  })),
+  queries: BatchQueryParams[],
+  client: SecretNetworkClient
+}) => sendSecretClientContractQuery$({
+  queryMsg: msgBatchQuery(queries),
+  client,
+  contractAddress,
+  codeHash,
+}).pipe(
   map((response) => parseBatchQuery(response as BatchQueryResponse)),
   first(),
+  catchError((err) => {
+    if (err.message.includes('{wasm contract}')) {
+      throw new Error('{wasm contract} error that typically occurs when batch size is too large and node gas query limits are exceeded. Consider reducing the batch size.');
+    } else {
+      throw new Error(err);
+    }
+  }),
 );
 
 /**
  * batch query of multiple contracts/message at a time
+ * @param batchSize defaults to processing all queries in a single batch
+ * when the batchSize is not passed in.
+ */
+const batchQuery$ = ({
+  contractAddress,
+  codeHash,
+  lcdEndpoint,
+  chainId,
+  queries,
+  batchSize,
+}:{
+  contractAddress: string,
+  codeHash?: string,
+  lcdEndpoint?: string,
+  chainId?: string,
+  queries: BatchQueryParams[],
+  batchSize?: number,
+}) => {
+  // if batch size is passed in, convert single batch into multiple batches,
+  // otherwise process all data in a single batch
+  const batches = batchSize
+    ? divideSingleBatchIntoArrayOfMultipleBatches(queries, batchSize)
+    : [queries]; // array of arrays required for the forkJoin
+
+  return getActiveQueryClient$(lcdEndpoint, chainId).pipe(
```

Comment:
Does this return the same type as a regular batch query? I know I've ran into issues of using fork join with flattening

Reply:
yes it should, my unit test proves that out I believe.